oak_attestation_verification/src/amd.rs

```diff
@@ -0,0 +1,199 @@
+}
+
+// Currently unused, use `pub` only to disable the warning.
+pub fn product_name(cert: &Certificate) -> anyhow::Result<String> {
+    let exts = cert
+        .tbs_certificate
+        .extensions
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("could not get extensions from cert"))?;
+    let pn_ext = exts
+        .iter()
+        .find(|&ext| ext.extn_id == PRODUCT_NAME_OID)
+        .ok_or_else(|| anyhow::anyhow!("no product name found in cert"))?;
+    String::from_utf8(pn_ext.extn_value.as_bytes().to_vec())
+        .map_err(|_utf8_err| anyhow::anyhow!("failed to read product name"))
+}
+
+fn chip_id(cert: &Certificate) -> anyhow::Result<[u8; 64]> {
+    let exts = cert
+        .tbs_certificate
+        .extensions
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("could not get extensions from cert"))?;
+    let chip_id_ext = exts
+        .iter()
+        .find(|&ext| ext.extn_id == CHIP_ID_OID)
+        .ok_or_else(|| anyhow::anyhow!("no chip ID found in cert"))?;
+    let chip_id = chip_id_ext.extn_value.as_bytes().to_vec();
+    anyhow::ensure!(chip_id.len() == 64, "length of chip ID value is not 64");
+
+    // Copy into array of fixed length.
+    let mut result = [0; 64];
+    result.copy_from_slice(&chip_id[0..64]);
+    Ok(result)
+}
+
+fn tcb_version(vcek: &Certificate) -> anyhow::Result<TcbVersion> {
+    let mut tcb = TcbVersion::new_zeroed();
+    for ext in vcek
+        .tbs_certificate
+        .extensions
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("could not get extensions from VCEK cert"))?
+    {
+        // What's the appropriate way of extracting u8 from OctetString?
+        let arr = ext.extn_value.as_bytes();
+        let last = arr.len() - 1;
+        if ext.extn_id == BL_OID {
+            tcb.boot_loader = arr[last];
+        } else if ext.extn_id == TEE_OID {
+            tcb.tee = arr[last];
+        } else if ext.extn_id == SNP_OID {
+            tcb.snp = arr[last];
+        } else if ext.extn_id == UCODE_OID {
+            tcb.microcode = arr[last];
+        }
+    }
+    Ok(tcb)
+}
+
+pub fn verify_attestation_report_signature(
+    vcek: &Certificate,
+    report: &AttestationReport,
+) -> anyhow::Result<()> {
+    // First check some necessary conditions for the signature to be valid.
+    let arpt_chip_id = report.data.chip_id;
+    let vcek_chip_id = chip_id(vcek)?;
+    anyhow::ensure!(
+        arpt_chip_id == vcek_chip_id,
+        "chip id differs attestation={} vcek={}",
+        hex::encode(arpt_chip_id),
+        hex::encode(vcek_chip_id)
+    );
+
+    let arpt_tcb = &report.data.reported_tcb;
+    let vcek_tcb = tcb_version(vcek)?;
+    anyhow::ensure!(
+        arpt_tcb.snp == vcek_tcb.snp
+            && arpt_tcb.microcode == vcek_tcb.microcode
+            && arpt_tcb.tee == vcek_tcb.tee
+            && arpt_tcb.boot_loader == vcek_tcb.boot_loader,
+        "mismatch in TCB version"
```

Comment:
Perhaps better to split into individual checks so that the error is clearer? Also it may be worth returning expected and actual values in the error message. Think how it would help ourselves in the future trying to debug something going wrong with a new version of the TCB.

Reply:
Done. Will look more into TCB versions with attestation report reference values. We need a formatter for the TCB version.